encoder.go

```diff
@@ -38,67 +46,93 @@ func newMaxEncoder() *encoder {
 	}
 }
 
+// Encode takes a Go value and returns a wafObject pointer and an error.
+// The returned wafObject is the root of the tree of imbriqued wafObjects representing the Go value.
+// The only error case is if the top-level object is "Unusable" which means that the data is nil or a non-data type
+// like a function or a channel.
 func (encoder *encoder) Encode(data any) (*wafObject, error) {
 	value := reflect.ValueOf(data)
 	wo := &wafObject{}
 
-	if err := encoder.encode(value, wo, encoder.objectMaxDepth); err != nil {
-		return nil, err
+	encoder.encode(value, wo, encoder.objectMaxDepth)
+
+	if wo._type == wafInvalidType {
+		return nil, errUnsupportedValue
 	}
 
 	return wo, nil
 }
 
-func encodeNative[T native](val T, t wafObjectType, obj *wafObject) error {
+func encodeNative[T native](val T, t wafObjectType, obj *wafObject) {
 	obj._type = t
```

Comment:
Seems to me that `obj._type` could be inferred from `val.(type)` here? The current API seems susceptible to subtle mis-use.

Reply:
I can't because of invalid and nil types